subsys/net/lib/mqtt_socket/mqtt_transport_socket_tcp.c

```diff
@@ -0,0 +1,124 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/** @file mqtt_transport_socket_tcp.h
+ *
+ * @brief Internal functions to handle transport over TCP socket.
+ */
+
+#define LOG_MODULE_NAME net_mqtt_sock_tcp
+#define NET_LOG_LEVEL CONFIG_MQTT_LOG_LEVEL
+
+#include <errno.h>
+#include <net/socket.h>
+#include <net/mqtt.h>
+
+#include "mqtt_os.h"
+
+/**@brief Handles connect request for TCP socket transport.
+ *
+ * @param[in] client Identifies the client on which the procedure is requested.
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ */
+int mqtt_client_tcp_connect(struct mqtt_client *client)
+{
+	const struct sockaddr *broker = client->broker;
+	int ret;
+
+	client->transport.tcp.sock = socket(broker->sa_family, SOCK_STREAM,
+					    IPPROTO_TCP);
+	if (client->transport.tcp.sock < 0) {
+		return -errno;
+	}
+
+	MQTT_TRC("Created socket %d", client->transport.tcp.sock);
+
+	size_t peer_addr_size = sizeof(struct sockaddr_in6);
+
+	if (broker->sa_family == AF_INET) {
+		peer_addr_size = sizeof(struct sockaddr_in);
+	}
+
+	ret = connect(client->transport.tcp.sock, client->broker,
+		      peer_addr_size);
+	if (ret < 0) {
+		(void)close(client->transport.tcp.sock);
+		return -errno;
+	}
+
+	MQTT_TRC("Connect completed");
+	return 0;
+}
+
+/**@brief Handles write requests on TCP socket transport.
+ *
+ * @param[in] client Identifies the client on which the procedure is requested.
+ * @param[in] data Data to be written on the transport.
+ * @param[in] datalen Length of data to be written on the transport.
+ *
+ * @retval 0 or an error code indicating reason for failure.
+ */
+int mqtt_client_tcp_write(struct mqtt_client *client, u8_t const *data,
+			  u32_t datalen)
+{
+	u32_t offset = 0;
+	int ret;
+
+	while (offset < datalen) {
+		ret = send(client->transport.tcp.sock, data + offset,
```

Comment:
Oh, so you used "non-blocking recv, blocking send" trick. Cute. Actually, can't even find a fault - the send may take longer than ping interval, but send is *activity*, so there's not need to do anything during send. Of course, individual chunks of send may take longer than ping interval, but then we're really slow and server should and will disconnect us.

But you should document explicitly that publish() and all other client->server operations are blocking, so user can't have more than one MQTT client per thread (should use multiple threads for that).

Reply:
I had an impression that non-blocking  sending on sockets is not fully supported in Zephyr?

Reply:
Nope, no specific limitations are known. There's https://github.com/zephyrproject-rtos/zephyr/issues/5842, but of course, that's a little bit different than "non-blocking sending on sockets".

But non-blocking send() is not going to help you. Suppose, you were called for publish(), and had a partial send(). What do you do? You either have to return to caller, but then app can ask to publish again, but you didn't even finished with previous! Or you could do your own sub-event loop, to keep recving incoming msg, but then you call app back, and it can ask you to send, again. That's that recursive calling back which I mentioned, which we can't allow, so we'll be inventing ways to preclude app to publish/etc while another send is in progress (why I asked about mutex), or queue them.

All is worse than what there's now. The current impl is smart. Well, it's dumb, and *that's* smart ;-). Exactly what we need for a small RTOS. Just need to deal with some omissions in handling and generalize API.